tests/test_writer.py

```diff
@@ -176,10 +172,46 @@ def test_no_validation_non_unique_columns(self, tmp_path):
         write_tfs(tmp_path / "temporary.tfs", df, validate=False)
         assert (tmp_path / "temporary.tfs").is_file()
 
+    def test_write_no_headers_dataframe(self, tmp_path, _pd_dataframe):
+        # We make sure providing a df without headers (a pd.DataFrame for
+        # instance) still writes a valid TFS file to disk. This is NOT
+        # the same as having empty headers (empty dict)!
+        df = _pd_dataframe
+        write_tfs(tmp_path / "temporary.tfs", df, validate=False)  # need to discard validation
+        new = read_tfs(tmp_path / "temporary.tfs")
+        assert_frame_equal(df, new, check_frame_type=False)  # since Dataframe and TfsDataFrame are different df types
+
+    # ----- Below are tests for files with MAD-NG features ----- #
+
+    def test_tfs_write_read_with_booleans(self, _tfs_dataframe_booleans, tmp_path):
+        write_location = tmp_path / "test.tfs"
+        write_tfs(write_location, _tfs_dataframe_booleans, validate="madng")  # booleans are MAD-NG feature
+        assert write_location.is_file()
+
+        new = read_tfs(write_location)
+        assert_tfs_frame_equal(_tfs_dataframe_booleans, new, check_exact=False)  # float precision can be an issue
+
+    def test_tfs_write_read_with_complex(self, _tfs_dataframe_complex, tmp_path):
+        write_location = tmp_path / "test.tfs"
+        write_tfs(write_location, _tfs_dataframe_complex, validate="madng")  # booleans are MAD-NG feature
+        assert write_location.is_file()
+
+        new = read_tfs(write_location)
+        assert_tfs_frame_equal(_tfs_dataframe_complex, new, check_exact=False)  # float precision can be an issue
+
+    def test_tfs_write_read_madng_like(self, _tfs_dataframe_madng, tmp_path):
+        write_location = tmp_path / "test.tfs"
+        write_tfs(write_location, _tfs_dataframe_madng, validate="madng")  # booleans are MAD-NG feature
+        assert write_location.is_file()
+
+        new = read_tfs(write_location)
+        assert_tfs_frame_equal(_tfs_dataframe_madng, new, check_exact=False)  # float precision can be an issue
```

Comment:
**FEATURE**

These are typical writer tests but use files for MAD-NG features.